### DISCUSSION TITLE
SLEIGH varnode_symbol is wrong type in goto

### DISCUSSION THREAD

**Discussion by c64cryptoboy** [NationalSecurityAgency/ghidra#4851] [Q&A] [answered]
Got a question:

I'm attempting to change how the 6502 indirect JMP works (to support the 6502 indirect JMP bug).
I've modified the existing 6502.slaspec indirect JMP in prep for my changes, and this works the same (I've tested it in pcode emulation):

```
ADDRI2: (imm16)   is imm16    { export *:2 imm16; }

:JMP ADDRI2     is (op=0x6c); ADDRI2
{
    goto [ADDRI2];
}
```

Q: What I want is

```
:JMP ADDRI2     is (op=0x6c); ADDRI2
{
    dest_addr = logic that computes the dest address that supports the 6502 indirect jmp bug;
    goto dest_addr;  # no [] dereferencing here -- already done in the dest_addr assignment logic
}
```
    
but I keep running into _"varnode_symbol dest_addr is wrong type (should be start, end, or operand) in goto destination"_? Toy examples:

This creates that compiler error:

```
:JMP ADDRI2     is (op=0x6c); ADDRI2
{
    local dest_addr:2 = 0x8000;
    goto dest_addr;
}
```

But this doesn't:

```
:JMP ADDRI2     is (op=0x6c); ADDRI2
{
    local dest_addr:2 = 0x8000;
    goto 0x8000;
}
```

This creates that compiler error:

```
:JMP ADDRI2     is (op=0x6c); ADDRI2
{
    local dest_addr = ADDRI2 - 2;
    goto dest_addr;
}
```

But this doesn't:

```
:JMP ADDRI2     is (op=0x6c); ADDRI2
{
    local dest_addr = ADDRI2 - 2;
    goto [dest_addr];
}
```

More context if needed (these are unchanged from the 6502.slaspec):

```
define token opbyte (8)
   op       = (0,7)
   
   aaa      = (5,7)
   bbb      = (2,4)
   cc       = (0,1)
;

define token data (16)
	imm16 = (0,15)
;
```

**Comment by GhidorahRex** [comment, marked as answer]
The brackets aren't dereferencing. It is, but not in the way you think. `goto varnode` means _goto that varnode_. Not goto the value of the varnode. If you want to goto the value within the varnode then you need `goto [varnode]`.

In your examples, `dest_addr` is a local unique varnode. It's part of the `unique` address space. You can't jump to a varnode in the `unique` address space. If you have a local variable, you need the brackets around it in the `goto`. It doesn't matter what calculations you have in `dest_addr = <some calculations here>;` you will need the brackets.

`ADDRI` (and `ADDRI2`) are subconstructors which can export any varnode - so they're not limited to local uniques, and they're not varnodes themselves - they represent a varnode. `export *:2 imm16;` for instance is exporting a varnode that is an address in the default address space (for the 6502 this space is `RAM`). You don't need the brackets because `ADDRI` is not a local variable in this instance, it is an address in the `RAM` space.

**Reply by c64cryptoboy** [reply]
Happy holidays, and thank you for the detailed response; having that clarification reduces my syntax search space :D

I do have a followup question on constructing a goto address.  The revised instruction logic should look something like this:

```
:JMP ADDRI2     is (op=0x6c); ADDRI2
{
    # support wrap-around bug for indirect JMP(ADDR) when ADDR low byte is 0xff
    local addr_for_hi_byte:2 = (ADDRI2 & 0xff00) | ((ADDRI2 + 1) & 0xff);
    local addr_for_lo_byte:2 = ADDRI2;
    local dest_hi:1 = *[RAM] addr_for_hi_byte;
    local dest_lo:1 = *[RAM] addr_for_lo_byte;
    local dest_addr = (dest_hi << 8) | dest_lo;
    goto [dest_addr];
}
```

In pcode emulation, I get the wrong program counter result, and the Ghidra GUI shows the following pcode:

```
8010 6c  ff  80       JMP        (DAT_80ff )           = ADh
                  $U4d00 :2 =  INT_AND  *[RAM ]0x80ff :2, 0xff00 :2
                  $U4d80 :2 =  INT_ADD  *[RAM ]0x80ff :2, 1:2
                  $U4e00 :2 =  INT_AND  $U4d80 :2, 0xff :2
                  $U4f00 :2 =  INT_OR  $U4d00 :2, $U4e00 :2
                  $U4f80 :2 =  COPY  *[RAM ]0x80ff :2
                  $U5080 :1 =  LOAD  RAM ($U4f00 :2)
                  $U5180 :1 =  LOAD  RAM ($U4f80 :2)
                  $U5200 :1 =  INT_LEFT  $U5080 :1, 8:4
                  $U5300 :1 =  INT_OR  $U5200 :1, $U5180 :1
                  BRANCHIND  $U5300 :1
```

The (first?) problem that sticks out to me is that the return result is size :1, not :2.
(Note: in my testing, hardcoded values of size 2 work just fine when emulated, such as "local dest_addr:2 = 0xFACE; goto [dest_addr];")

So I'm trying to figure out how to make dest_addr be size :2.

If I try:
`local dest_addr:2 = ((dest_hi << 8) | dest_lo)`
The compiler complains that there's a "Problem in 'Or(|)' operator" and that "the output and all input sizes must match"

I gave the ellipsis operator a shot, but looks like it's not applicable here?
`local dest_addr:2 = ((dest_hi << 8) | dest_lo) ... & 0xffff;`
as the compiler is not expecting it

**Reply by GhidorahRex** [reply]
The ellipsis is only applicable in the bitpattern section (`is (oc=0x6c); ADDRI2`), not the pcode section.

The error is telling you that for an `or` operator you need the varnode sizes of both inputs and the output to all be the same. Since you didn't originally specify a size, the local variable `dest_addr` assumed it was the same as the input sizes, which is not what you want here, so you specifically give it a size.

But in order for the `or` to work, you need to convert the sizes of the two input varnodes to be the same size as the output. Perhaps the easiest way would be to extend the values. For logic operators like `or` you want to zero extend.

The correct fix would be:
`local dest_addr:2 = (zext(dest_hi) << 8 )| zext(dest_lo);`

**Reply by c64cryptoboy** [reply]
Thanks again for the fix!  One step closer to figuring this out...
